### DISCUSSION TITLE
SimilaritymapsFromModel

### DISCUSSION THREAD

**Discussion by PARODBE** [rdkit/rdkit#5267] [Q&A]
Hi!

I have a question, i am drawing the similaritymaps from model, I am going with an example:

![image](https://user-images.githubusercontent.com/62376157/167129862-3788980f-b8d1-4dd0-ac1c-27900c0cd2a1.png)

I have a classification problem, for this reason for similaritymaps I am using predict_proba, and I obtain the next result:

![image](https://user-images.githubusercontent.com/62376157/167130004-1bf62d9e-4be8-43fb-8cd7-5baaae38fd42.png)

Next, if I draw the similarity maps from model:

![image](https://user-images.githubusercontent.com/62376157/167130104-e8674547-5cc3-4284-a632-6718ab84dbff.png)

In my opinion, If you see the last picture, what is the reason for don't have the atom 2 with a delta value equal than atom with index 1 since the intensity of color is the same???

Another question when probability is lower than 0.5 the colour is blue? and higgher than 0.5 red?

**Comment by greglandrum** [comment]
Hi @PARODBE, there's not enough information here for us to be able to help you.
Where did the data in your table come from? How did you generate the similarity map?

**Comment by PARODBE** [comment]
Hi @greglandrum , first at all for the table with delta values I did this: 

from rdkit.Chem.Draw import SimilarityMaps

res = []
f=[]
for atm in example_mol.GetAtoms():
    idx = atm.GetIdx()
    fp = SimilarityMaps.GetMorganFingerprint(example_mol,idx)
    pred_val = model.predict_proba(np.array([fp]))[0][1]
    delta = example_pred - pred_val
    res.append([atm.GetSymbol(),idx,pred_val,delta])
    f.append(fp)
tmp_df = pd.DataFrame(res,columns = ["Atom Type","Atom Index","Predicted Value","Delta"])
tmp_df.sort_values("Delta",ascending=False)


And for the similarity maps from model:

def show_png(data):
    bio = io.BytesIO(data)
    img = Image.open(bio)
    return img

def getProba(fp, predctionFunction):
    return predctionFunction((fp,))[0][1]

def plot_similarity_map(mol, model):
    d = Draw.MolDraw2DCairo(1,1)
    SimilarityMaps.GetSimilarityMapForModel(mol,partial(SimilarityMaps.GetMorganFingerprint, nBits=2048, radius=2,useFeatures=True),
                                            lambda x: getProba(x, model.predict_proba),colorMap='coolwarm',size=(200,200))
    d.FinishDrawing()
    return d

from rdkit.Chem import Draw
from functools import partial

test_row = 3
test_mol = test_mec_medio.ROMol.values[test_row]
res = plot_similarity_map(test_mol,model);
show_png(res.GetDrawingText())

**Reply by swapchavan** [reply]
Hi Pablo @PARODBE , When I used "GetSimilarityMapFromWeights" module, I got correct image output!

```
temp_smi = "NCC1(CC(O)=O)CCCCC1"
temp_mol = Chem.MolFromSmiles(temp_smi)
for atom in temp_mol.GetAtoms():
    atom.SetAtomMapNum(atom.GetIdx())
temp_mol
wt_vec = [-0.030651,-0.030651,-0.030651,-0.030651,-0.030651,-0.036145,-0.054493,-0.054493,-0.061606,-0.061606,-0.061606,-0.061606]
SimilarityMaps.GetSimilarityMapFromWeights(temp_mol, wt_vec)
```


I think the delta's in the result table shown above may not be equivalent to delta's you are getting using "GetSimilarityMapForModel" module.
Could you not manually obtain delta for your query smiles? E.g. for 

```
smi = "NCC1(CC(O)=O)CCCCC1"
m = Chem.MolFromSmiles(smi)
edit_m = Chem.RWMol(m)
# Replace 12th atom by Yttrium and then replace Yttrium by ''
edit_m.GetAtomWithIdx(11).SetAtomicNum(39)
masked_smiles = Chem.MolToSmiles(edit_m).replace('[Y]','')
```
And getting predict_proba for this masked_smiles, will give you delta for 12th atom. Repeat for all remaining 11 atoms.
Store all deltas in a list and use "GetSimilarityMapFromWeights" module to create image!
Good luck
/Swapnil

**Reply by PARODBE** [reply]
Hi Swapnil,

Thank you for your answer. And now all this has more sense for me. Only a question please, I was having an error with seting atomic number in 39 and I have set it to 30 and It is working for me. What number is more correct? Or this is not so important?

Thanks!
Pablo

**Reply by swapchavan** [reply]
Yttrium valency is +3, so your error must be becoz of that! There is no right answer which atomic number you should use. I use 39 becoz it is rare earth metal so very very tiny percent of chemicals will likely have it in their smiles. The imporant thing to remember is - we don't want to replace existing atom(s)! I mean imagin if you mutate character of your interest, by 7 (nitrogen), and then remove it. You will likely remove native nitrogens from that molecule. So use any atomic number that won't be present in any of your training set smiles.
